plugin/trino-hive/src/main/java/io/trino/plugin/hive/procedure/SyncPartitionMetadataProcedure.java

```diff
@@ -137,23 +139,46 @@ private void doSyncPartitionMetadata(ConnectorSession session, ConnectorAccessCo
                 accessControl.checkCanDeleteFromTable(null, new SchemaTableName(schemaName, tableName));
             }
 
-            Location tableLocation = Location.of(table.getStorage().getLocation());
-
-            Set<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName)
+            Set<String> canonicalPartitionNamesInMetastore = metastore.getPartitionNames(schemaName, tableName)
                     .map(ImmutableSet::copyOf)
                     .orElseThrow(() -> new TableNotFoundException(schemaTableName));
-            Set<String> partitionsInFileSystem = listPartitions(fileSystemFactory.create(session), tableLocation, table.getPartitionColumns(), caseSensitive);
+            String tableStorageLocation = table.getStorage().getLocation();
+            String tableLocationDirectory = tableStorageLocation.endsWith("/") ? tableStorageLocation : tableStorageLocation + "/";
+            Set<String> actualPartitionNamesInMetastore = metastore.getPartitionsByNames(schemaName, tableName, canonicalPartitionNamesInMetastore.stream().collect(toImmutableList()))
+                    .values().stream()
+                    .filter(Optional::isPresent) // disregard partitions which disappeared in the meantime since listing the partition names
+                    .map(Optional::get)
+                    .map(partition -> getPartitionNameFromPartitionLocation(partition.getStorage().getLocation(), tableLocationDirectory))
+                    .collect(toImmutableSet());
+
+            Set<String> partitionsInFileSystem = listPartitions(fileSystemFactory.create(session), Location.of(tableStorageLocation), table.getPartitionColumns(), caseSensitive);
 
             // partitions in file system but not in metastore
-            Set<String> partitionsToAdd = difference(partitionsInFileSystem, partitionsInMetastore);
+            Set<String> partitionsToAdd = difference(partitionsInFileSystem, actualPartitionNamesInMetastore);
 
             // partitions in metastore but not in file system
-            Set<String> partitionsToDrop = difference(partitionsInMetastore, partitionsInFileSystem);
+            Set<String> partitionsToDrop = difference(actualPartitionNamesInMetastore, partitionsInFileSystem);
 
             syncPartitions(partitionsToAdd, partitionsToDrop, syncMode, metastore, session, table);
         }
     }
 
+    private static String getPartitionNameFromPartitionLocation(String partitionLocation, String tableLocation)
+    {
+        if (!partitionLocation.startsWith(tableLocation)) {
+            throw new IllegalArgumentException("Partition location '%s' does not belong to base table location '%s'".formatted(partitionLocation, tableLocation));
+        }
+
+        String partitionName = partitionLocation.substring(tableLocation.length());
```

Comment:
`partitionName` will start with a slash if `tableLocation` does not end in a slash. We should check for this here and remove the leading slash.